libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp

```diff
@@ -220,470 +233,658 @@ void AP_Frsky_Telem::sport_tick(void)
+ * for FrSky SPort Passthrough (OpenTX) protocol (X-receivers)
  */
-void AP_Frsky_Telem::send_batt_remain(void)
+uint32_t AP_Frsky_Telem::calc_batt(void)
 {
-    frsky_send_data(FRSKY_ID_FUEL, _batt_remaining);
-}
+    uint32_t batt = 0;
 
-/*
- * send battery voltage 
- */
-void AP_Frsky_Telem::send_batt_volts(void)
-{
-    frsky_send_data(FRSKY_ID_VFAS, _batt_volts);
+    batt |= (((uint16_t)roundf(_battery.voltage() * 10.0f)) & 0x1FF); // battery voltage in decivolts, can have up to a 12S battery (4.25Vx12S = 51.0V)
+    batt |= prep_number(roundf(_battery.current_amps() * 10.0f), 2, 1)<<9; // battery current draw in deciamps
+    // battery current draw since power on in mAh
+    if (_battery.current_total_mah() > 32767) { // limit to 32767 if above since value is encoded on 15 bits...
+        batt |= 0x7FFF<<17;
+    } else {
+        batt |= (((uint16_t)roundf(_battery.current_total_mah())) & 0x7FFF)<<17;
+    }
+    return batt;
 }
 
 /*
- * send current consumptiom 
+ * prepare various ardupilot status data
+ * for FrSky SPort Passthrough (OpenTX) protocol (X-receivers)
  */
-void AP_Frsky_Telem::send_current(void)
+uint32_t AP_Frsky_Telem::calc_ap_status(void)
 {
-    frsky_send_data(FRSKY_ID_CURRENT, _batt_amps);
-}
+    uint32_t ap_status = 0;
 
-/*
- * send heading in degree based on AHRS and not GPS 
- */
-void AP_Frsky_Telem::send_heading(void)
-{
-    frsky_send_data(FRSKY_ID_GPS_COURS_BP, _course_in_degrees);
+    ap_status |= (uint8_t)(*_ap.control_mode & 0x1F); // flight mode number
+    ap_status |= (uint8_t)(*_ap.value & 0x6)<<4; // simple/super simple modes flags
+    ap_status |= (uint8_t)(*_ap.value & 0x80); // land complete flag
+    ap_status |= (uint8_t)(AP_Notify::flags.armed)<<8; // armed flag
+    ap_status |= (uint8_t)(AP_Notify::flags.failsafe_battery)<<9; // battery failsafe flag
+    ap_status |= (uint8_t)(AP_Notify::flags.ekf_bad)<<10; // bad ekf flag
+    return ap_status;
 }
 
 /*
- * send gps latitude degree and minute integer part; Initialize gps info
+ * prepare home position related data
+ * for FrSky SPort Passthrough (OpenTX) protocol (X-receivers)
  */
-void AP_Frsky_Telem::send_gps_lat_dd(void)
+uint32_t AP_Frsky_Telem::calc_home(void)
 {
-    frsky_send_data(FRSKY_ID_GPS_LAT_BP, _latdddmm);
-}
+    uint32_t home = 0;
 
-/*
- * send gps latitude minutes decimal part 
- */
-void AP_Frsky_Telem::send_gps_lat_mm(void)
-{
-    frsky_send_data(FRSKY_ID_GPS_LAT_AP, _latmmmm);
+    // distance between vehicle and home location in meters
+    home |= prep_number(roundf(*_ap.home_distance * 0.01f), 3, 2);
+    // altitude between vehicle and home location in decimeters
+    Location current_loc;
+    if  (_ahrs.get_position(current_loc)) {
+        if (current_loc.flags.relative_alt) {
+            _current_height = current_loc.alt;
+        } else {
+            _current_height = current_loc.alt - _ahrs.get_home().alt;
+        }
+    }
+    home |= prep_number(roundf(_current_height * 0.1f), 3, 2)<<12;
```

Comment:
if _ahrs.get_position(current_loc) is false, then _current_height == 0. Is that handled correctly here?